### PR TITLE
Improve trace reconciliation performance

### PR DESCRIPTION
**Pull request by bendikberg**
### Purpose

When loading a file with a lot of bindings stored in CallSite.SingleRunTraceData, the trace reconciliation spends a lot of time reading nested data due to list allocations. And in general it does unecessary work copying lists and iterating through IEnumerables, for .Contains-testing and the like.

These changes have a large impact on the graph execution time for graphs with a lot of orphaned serializable data, as seen in these profiling results.

devenv_Hu6ocw92K8
devenv_qPyrZ0YqIQ

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [X] This PR contains no files larger than 50 MB

### Release Notes

N/A

### Reviewers

(FILL ME IN) Reviewer 1 (If possible, assign the Reviewer for the PR)
@mjkkirschner 

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of
@dimven

### CODE REVIEW

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 1]
```diff
@@ -472,13 +475,11 @@ private static string CompressSerializedTraceData(string json)
...
-            var result = new List<string>();
-
-            if (!beforeFirstRunSerializables.Any())
-                return result;
+            if (beforeFirstRunSerializables.Count == 0)
+                return new List<string>();
 
-            var currentSerializables = traceData.SelectMany(td => td.RecursiveGetNestedData());
-            result.AddRange(beforeFirstRunSerializables.Where(hs => !currentSerializables.Contains(hs)).ToList());
+            var currentSerializables = traceData.SelectMany(td => td.RecursiveGetNestedData()).ToHashSet();
```
Comment (bendikberg) [inline src/Engine/ProtoCore/Lang/CallSite.cs:481]: This line with the added .ToHashSet() call and the line below are the likeliest culprits of the bad performance, as they would previously keep reloading the TraceData from the CallSite object for every .Contains test

File: src/DynamoCore/Engine/EngineController.cs [hunk 2]
```diff
@@ -533,14 +533,14 @@ internal void ReconcileTraceDataAndNotify()
             var callsiteToOrphanMap = new Dictionary<Guid, List<string>>();
             foreach (var cs in liveRunnerServices.RuntimeCore.RuntimeData.CallsiteCache.Values)
             {
-                var orphanedSerializables = cs.GetOrphanedSerializables().ToList();
-                if (callsiteToOrphanMap.ContainsKey(cs.CallSiteID))
+                var orphanedSerializables = cs.GetOrphanedSerializables();
+                if (callsiteToOrphanMap.TryGetValue(cs.CallSiteID, out var serializablesForCallsite))
```
Comment (mjkkirschner) [inline src/DynamoCore/Engine/EngineController.cs:537]: how does this change improve performance?

File: src/DynamoCore/Engine/EngineController.cs [hunk 3]
```diff
@@ -533,14 +533,14 @@ internal void ReconcileTraceDataAndNotify()
...
                 {
-                    callsiteToOrphanMap[cs.CallSiteID].AddRange(orphanedSerializables);
+                    serializablesForCallsite.AddRange(orphanedSerializables);
                 }
                 else
                 {
-                    callsiteToOrphanMap.Add(cs.CallSiteID, orphanedSerializables);
+                    callsiteToOrphanMap.Add(cs.CallSiteID, orphanedSerializables.ToList());
```
Comment (mjkkirschner) [inline src/DynamoCore/Engine/EngineController.cs:543]: so `orphanedSerializables` is already a `List<string>` - https://github.com/DynamoDS/Dynamo/blob/10df94ff5d0542328aee7205d6259aa0fb4f4921/src/Engine/ProtoCore/Lang/CallSite.cs#L475 though it's declared here as an `IList` - I assume that this calls the List constructor passing in the existing list, have you verified there is any benefit to moving this `ToList` call around?

If it's really such a big performance impact you could probably change this data structure to be a dict of `IList`

File: src/DynamoCore/Graph/Workspaces/HomeWorkspaceModel.cs [hunk 4]
```diff
@@ -897,7 +897,7 @@ internal IList<string> GetOrphanedSerializablesAndClearHistoricalTraceData()
 
                 if (Nodes.All(n => n.GUID != nodeGuid))
                 {
-                    orphans.AddRange(nodeData.Value.SelectMany(CallSite.GetAllSerializablesFromSingleRunTraceData).ToList());
+                    orphans.AddRange(nodeData.Value.SelectMany(CallSite.GetAllSerializablesFromSingleRunTraceData));
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/HomeWorkspaceModel.cs:900]: does calling add range immediately execute the deferred selectMany call?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 5]
```diff
@@ -1260,22 +1260,30 @@ private void EngineController_TraceReconcliationComplete(TraceReconciliationEven
             // This dictionary gets redistributed into a dictionary keyed by the workspace id.
 
             var workspaceOrphanMap = new Dictionary<Guid, List<string>>();
+            var nodeToWorkspaceMap = new Dictionary<Guid, WorkspaceModel>();
 
-            foreach (var ws in Workspaces.OfType<HomeWorkspaceModel>())
+            foreach (var maybeWs in Workspaces)
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:1265]: why do this?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 6]
```diff
@@ -1260,22 +1260,30 @@ private void EngineController_TraceReconcliationComplete(TraceReconciliationEven
...
             {
+                foreach (var node in maybeWs.Nodes)
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:1267]: please use braces.

Comment (mjkkirschner) [reply src/DynamoCore/Models/DynamoModel.cs:1267]: also consider that this could be a very large number of nodes.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 7]
```diff
@@ -1287,23 +1295,26 @@ private void EngineController_TraceReconcliationComplete(TraceReconciliationEven
...
                 // Find the owning workspace for a node.
-                var nodeSpace =
-                    Workspaces.FirstOrDefault(
-                        ws =>
-                            ws.Nodes.FirstOrDefault(n => n.GUID == nodeGuid)
-                                != null);
+                if (!nodeToWorkspaceMap.TryGetValue(nodeGuid, out var nodeSpace))
+                    continue;
+
+                //var nodeSpace =
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:1301]: ??

File: src/DynamoCore/Models/DynamoModel.cs (hunk 5 above)
Comment (mjkkirschner) [reply src/DynamoCore/Models/DynamoModel.cs:1265]: I don't think there is any value in iterating all the nodes of a custom node workspace... maybe you can say a bit more about why you need this other map to improve performance.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 8]
```diff
@@ -1260,22 +1260,30 @@ private void EngineController_TraceReconcliationComplete(TraceReconciliationEven
...
+                    continue;
+
                 // Get the orphaned serializables to this workspace
-                var wsOrphans = ws.GetOrphanedSerializablesAndClearHistoricalTraceData().ToList();
+                var wsOrphans = (List<string>)ws.GetOrphanedSerializablesAndClearHistoricalTraceData();
 
                 if (!wsOrphans.Any())
                     continue;
 
-                if (!workspaceOrphanMap.ContainsKey(ws.Guid))
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:1272]: whats the benefit of this change?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 9]
```diff
@@ -1287,23 +1295,26 @@ private void EngineController_TraceReconcliationComplete(TraceReconciliationEven
 
                 // TODO: MAGN-7314
                 // Find the owning workspace for a node.
-                var nodeSpace =
-                    Workspaces.FirstOrDefault(
-                        ws =>
-                            ws.Nodes.FirstOrDefault(n => n.GUID == nodeGuid)
-                                != null);
+                if (!nodeToWorkspaceMap.TryGetValue(nodeGuid, out var nodeSpace))
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:1298]: hmm... on first glance memoizing this lookup makes sense, but I'm unclear exactly how many times we recreate the map vs how fast we bail on first finding the node in the existing code -can you look into that?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 10]
```diff
@@ -1287,23 +1295,26 @@ private void EngineController_TraceReconcliationComplete(TraceReconciliationEven
...
+                //        ws =>
+                //            ws.Nodes.FirstOrDefault(n => n.GUID == nodeGuid)
+                //                != null);
 
-                if (nodeSpace == null) continue;
+                //if (nodeSpace == null) continue;
 
                 // Add the node's orphaned serializables to the workspace
                 // orphan map.
-                if (workspaceOrphanMap.ContainsKey(nodeSpace.Guid))
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:1300]: same question about this lookup change.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 11]
```diff
@@ -155,17 +155,20 @@ public bool Contains(string data)
             public List<string> RecursiveGetNestedData()
             {
                 List<string> ret = new List<string>();
+                RecursiveFillWithNestedData(ret);
+                return ret;
+            }
 
+            public void RecursiveFillWithNestedData(List<string> listToFill)
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/CallSite.cs:162]: seems this can be private or internal.

Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/CallSite.cs:162]: what is the intention of your change here?

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 1 above)
Comment (mjkkirschner) [reply src/Engine/ProtoCore/Lang/CallSite.cs:481]: if I'm understanding that correctly, then it could be solved by actualizing the `currentSerializables` outside the deferred call - right?

Is the cost to create the hashset worth the lookup cost?

**Comment by mjkkirschner** [comment]
this did pass here once:
https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16982/

File: src/DynamoCore/Engine/EngineController.cs (hunk 2 above)
Comment (bendikberg) [reply src/DynamoCore/Engine/EngineController.cs:537]: Since the immediate action after the ContainsKey is to mutate the value at that key, using TryGetValue avoids a redundant dictionary lookup

[C# lint CA1854](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1854)

**Comment by dimven** [comment]
@mjkkirschner This is more relevant to D4R because that writes huge amounts of trace data tying in the dynamo wrapper elements to the matching Revit elements for continuous updates between dynamo sessions. In cases where the graph stored a lot of references that are no longer valid, the first execution would be very slow. Look at the attached trace data images. The `GetOrphanedSerializables` method was taking 42% of the graph execution time and after the changes that dropped to an insignificant number.

I'm not familiar with Dynamo for Civil3d but if it tracks the elements in a similar way, then this would be relevant there as well.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 12]
```diff
@@ -472,13 +475,11 @@ private static string CompressSerializedTraceData(string json)
         /// </summary>
         public IList<string> GetOrphanedSerializables()
         {
-            var result = new List<string>();
-
-            if (!beforeFirstRunSerializables.Any())
-                return result;
+            if (beforeFirstRunSerializables.Count == 0)
+                return new List<string>();
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs:479]: Is this any more performant?

File: src/DynamoCore/Engine/EngineController.cs (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCore/Engine/EngineController.cs:537]: okay, seems like a pretty safe change, though under profiling does it make an actual difference here?

**Comment by bendikberg** [comment]
@mjkkirschner

> hey @bendikberg I would like some more description in the PR about what your changes are intending and how you validated that they make a substantial improvement. Maybe not all of these changes are equally important.

I made this PR during the D4R performance profiling project. This code was originally written to be more of a research and test thing, and is in a pretty raw state. The reason for the raw state of the PR is that I assumed someone on the Dynamo team would just grab the parts of this PR they deemed relevant and implement those.

The thing I was trying to do was to figure out where and why the Callsite TraceData was slowing down the first run of a graph, so I touched on a few different files related to the TraceData handling.

Most of the .ToList and ContainsKey/TryGetValue-juggling can be left out of this PR or skipped entirely when they are compared to the performance gained from algorithmic optmizations. They don't cause the same type of exponentital slowdowns, but still do unecessary work. I made those changes since I figured as long as I was upgrading the TraceData performance on the whole, I would also include simple (but tiny) optimizations.


I made some benchmarks for the most impactful changes:

---

This gets slower as the number of TraceData items on a node increases (this can easily get very large for a Revit file).

[RecursiveFillWithNestedData change](https://github.com/DynamoDS/Dynamo/pull/15725/files#diff-862cf04aa1be46c6108d203e342593db6d58cd0cd03d6296b8e89290f38b7480R162)
```
Tests for 0 sublevels, 0 children per item per level (total: 1)
        RecursiveFillManyLists         :   100 runs : mean:         13 us
        RecursiveFillOneList           :   100 runs : mean:          4 us
        NonRecursiveFill               :   100 runs : mean:          4 us
Tests for 1 sublevels, 100 children per item per level (total: 101)
        RecursiveFillManyLists         :   100 runs : mean:         32 us
        RecursiveFillOneList           :   100 runs : mean:         21 us
        NonRecursiveFill               :   100 runs : mean:         22 us
Tests for 100 sublevels, 1 children per item per level (total: 101)
        RecursiveFillManyLists         :   100 runs : mean:         48 us
        RecursiveFillOneList           :   100 runs : mean:         22 us
        NonRecursiveFill               :   100 runs : mean:         19 us
Tests for 1 sublevels, 10000 children per item per level (total: 10001)
        RecursiveFillManyLists         :   100 runs : mean:        656 us
        RecursiveFillOneList           :   100 runs : mean:        364 us
        NonRecursiveFill               :   100 runs : mean:        450 us
Tests for 1 sublevels, 100000 children per item per level (total: 100001)
        RecursiveFillManyLists         :   100 runs : mean:      15818 us
        RecursiveFillOneList           :   100 runs : mean:       2122 us
        NonRecursiveFill               :   100 runs : mean:       3612 us
```

---

This scales poorly with the number of nodes on a canvas.

[Node memoization](https://github.com/DynamoDS/Dynamo/pull/15725/files#diff-d730c0ae93cb1da6c95f0ba453dd94553067feddaedf354cb63a6bfcde71ffcfR1267)

[Node lookup](https://github.com/DynamoDS/Dynamo/pull/15725/files#diff-d730c0ae93cb1da6c95f0ba453dd94553067feddaedf354cb63a6bfcde71ffcfR1298)

```
Test input: Workspaces 1 with 10 nodes. Selected workspace: 1. Lookup: 1 node ids RandomFromWorkspace
        Lookup linq (current)          :   100 runs : mean:          2 us
        Lookup dict                    :   100 runs : mean:          3 us
Test input: Workspaces 1 with 10 nodes. Selected workspace: 1. Lookup: 5 node ids RandomFromWorkspace
        Lookup linq (current)          :   100 runs : mean:         12 us
        Lookup dict                    :   100 runs : mean:          7 us
Test input: Workspaces 1 with 100 nodes. Selected workspace: 1. Lookup: 100 node ids AllInWorkspace
        Lookup linq (current)          :   100 runs : mean:       1974 us
        Lookup dict                    :   100 runs : mean:         36 us
Test input: Workspaces 1 with 1000 nodes. Selected workspace: 1. Lookup: 1000 node ids AllInWorkspace
        Lookup linq (current)          :   100 runs : mean:     136409 us
        Lookup dict                    :   100 runs : mean:        221 us
Test input: Workspaces 1 with 1000 nodes. Selected workspace: 1. Lookup: 500 node ids RandomFromWorkspace
        Lookup linq (current)          :   100 runs : mean:      67649 us
        Lookup dict                    :   100 runs : mean:        218 us
Test input: Workspaces 1 with 1000 nodes. Selected workspace: 1. Lookup: 1000 node ids NotInWorkspace
        Lookup linq (current)          :   100 runs : mean:     130951 us
        Lookup dict                    :   100 runs : mean:        250 us
Test input: Workspaces 5 with 100 nodes. Selected workspace: 5. Lookup: 10 node ids RandomFromWorkspace
        Lookup linq (current)          :   100 runs : mean:        707 us
        Lookup dict                    :   100 runs : mean:         99 us
Test input: Workspaces 5 with 100 nodes. Selected workspace: 5. Lookup: 50 node ids RandomFromWorkspace
        Lookup linq (current)          :   100 runs : mean:       3382 us
        Lookup dict                    :   100 runs : mean:        108 us
Test input: Workspaces 5 with 100 nodes. Selected workspace: 5. Lookup: 1000 node ids NotInWorkspace
        Lookup linq (current)          :   100 runs : mean:      67577 us
        Lookup dict                    :   100 runs : mean:        143 us
```

---

[The main culprit](https://github.com/DynamoDS/Dynamo/pull/15725/files#diff-862cf04aa1be46c6108d203e342593db6d58cd0cd03d6296b8e89290f38b7480R481)
```
Tests for 10 callsites, each with 10 tracedata (datatype System.Guid)
        Lookup linq (current)          :    10 runs : mean:        658 us
        Lookup dict (fill many lists)  :    50 runs : mean:         65 us
        Lookup dict (fill one list)    :    50 runs : mean:         56 us
Tests for 10 callsites, each with 100 tracedata (datatype System.Guid)
        Lookup linq (current)          :    10 runs : mean:      13441 us
        Lookup dict (fill many lists)  :    50 runs : mean:        208 us
        Lookup dict (fill one list)    :    50 runs : mean:        231 us
Tests for 10 callsites, each with 1000 tracedata (datatype System.Guid)
        Lookup linq (current)          :    10 runs : mean:     426107 us
        Lookup dict (fill many lists)  :    50 runs : mean:       1589 us
        Lookup dict (fill one list)    :    50 runs : mean:       1185 us
Tests for 10 callsites, each with 10000 tracedata (datatype System.Guid)
        Lookup linq (current)          :     1 run        :  142486207 us
        Lookup dict (fill many lists)  :    50 runs : mean:      30172 us
        Lookup dict (fill one list)    :    50 runs : mean:      15650 us
```

---
[Benchmark source](https://gist.github.com/bendikberg/4c5254cc0c3a9f409f5026ef7ff05bc8)


For small input sizes it doesn't add up to many microseconds, but it scales very poorly with larger input, which is bad since users might save whatever data in their files, which will make Dynamo hang for seconds or minutes.

---

### TL;DR:

This PR can be reduced to only include the algorithmic improvements that I've shown in the benchmarks.

File: src/DynamoCore/Models/DynamoModel.cs (hunk 9 above)
Comment (mjkkirschner) [reply src/DynamoCore/Models/DynamoModel.cs:1298]: discussed on slack- seems the cost to create this map each run is likely worth it so the lookup is much faster later. I do still have question (asked above):

Do we need to put nodes from custom node workspaces into the map - I kind of remember that nodes in custom node workspaces do participate in trace, it's just that custom node workspaces don't serialize trace - maybe that does not matter here?

Can you look into this?

**Comment by mjkkirschner** [comment]
This has now passed on master-5 a bunch of times in a row, so thats great. I am thinking about how we might stick this behind a feature flag.

**Comment by mjkkirschner** [comment]
I gave up on getting this behind a feature flag as some of the core changes I wanted to guard are inside protocore - where getting the feature flags manager is harder and requires a refactor to do safely. I will file a followup for that.

In the meantime I will merge this after it passes on master-15.
FYI @jnealb

**Comment by mjkkirschner** [comment]
passed on master-15, the two failures in the latest master-5 run appear to be flaky looking at their history in jenkins.

**Review by mjkkirschner — APPROVED** [review]
